Plugins/BranchTree/BranchTreeForm.cs

```diff
@@ -0,0 +1,137 @@
+        }
+
+        public BranchTreeForm(GitUIBaseEventArgs gitUiCommands) : this()
+        {
+            _gitUiCommands = gitUiCommands.GitUICommands;
+            var gitModule = gitUiCommands != null ? gitUiCommands.GitModule : null;
+            
+            BranchesTreeView.BeginUpdate();
+            BranchesTreeView.Nodes.Clear();
+
+            _localTreeNode = BranchesTreeView.Nodes.Add("Local");
+            _remoteTreeNode = BranchesTreeView.Nodes.Add("Remote");
+            _tagTreeNode = BranchesTreeView.Nodes.Add("Tag");
+
+            InitLocalBranches(_localTreeNode, gitModule);
+            InitRemoteBranches(_remoteTreeNode, gitModule);
+            InitTag(_tagTreeNode, gitModule);
+          
+            BranchesTreeView.EndUpdate();
+            BranchesTreeView.NodeMouseDoubleClick += Branches_NodeMouseDoubleClick;
+        }
+
+        private void InitTag(TreeNode root, IGitModule _gitModule)
+        {
+            var tags = _gitModule.GetRefs(true, false)
+                .OrderBy(r => r.Name);
+
+            foreach (var tag in tags)
+            {
+                AddOrUpdateNode(root.Nodes, tag.Name.Split('/'), tag.CompleteName);
+            }
+        }
+
+        private void InitRemoteBranches(TreeNode root, IGitModule gitModule)
+        {
+            var branches = gitModule.GetRefs()
+                   .Where(branch => branch.IsRemote && !branch.IsTag)
+                   .OrderBy(r => r.Name);                   
+
+            var remotes = gitModule.GetRemotes(allowEmpty: true);
+            
+            foreach (var branch in branches)
+            {
+                var remote = branch.Name.Split('/').First();
+
+                if (!remotes.Contains(remote))
+                {
+                    continue;
+                }
+
+                AddOrUpdateNode(root.Nodes, branch.Name.Split('/'), branch.CompleteName);
+            }
+        }
+
+        private void InitLocalBranches(TreeNode root, IGitModule gitModule)
+        {
+            var localBranches = gitModule.RunGitCmd("branch")
+                .Split('\n')
+                .Where(branch => !string.IsNullOrWhiteSpace(branch))// first is ""
+                .OrderByDescending(branch => branch.Contains('*'))// * for current branch
+                .ThenBy(r => r)
+                .Select(line => line.Trim());// trim justify space
+            foreach (var localBranch in localBranches)
+            {
+                AddOrUpdateNode(root.Nodes, localBranch.Split('/'), localBranch);
+            }
+            
+        }
+                
+
+        private void Branches_NodeMouseDoubleClick(object sender, TreeNodeMouseClickEventArgs e)
+        {
+            if (e.Node.Tag != null)
+            {
+                MessageBox.Show("Quick switching coming soon!", "Sorry but", MessageBoxButtons.OK);
+            }
+        }
+
+        private void AddOrUpdateNode(TreeNodeCollection nodes, string[] children, string fullPath)
```

Comment:
I am unsure about this method, it look inefficient. We call this method in multiple places from loops and we making new memory allocations within the method.... 
Could you refactor the implementation so the method is invoked only once with a collection containing tags, locals and remotes and it makes a single iteration over the collection populating the tree?